File: easy/783. Minimum Distance Between BST Nodes/783. Minimum Distance Between BST Nodes.py
from typing import Optional, Tuple


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        def run(
            *, tree: Optional[TreeNode], minimum: int, prev: Optional[int]
        ) -> Tuple[Optional[int], int]:
            # if tree.left:
            #     minimum = min(minimum, tree.val - tree.left.val)

            # if tree.right:
            #     minimum = min(minimum, tree.right.val - tree.val)

            # return min(
            #     run(tree=tree.left, minimum=minimum),
            #     run(tree=tree.right, minimum=minimum),
            # )

            if not tree:
                return prev, minimum

            prev, minimum = run(tree=tree.left, minimum=minimum, prev=prev)

            if prev is not None:
                minimum = min(minimum, tree.val - prev)

            prev = tree.val

            return run(tree=tree.right, minimum=minimum, prev=prev)

        _, res = run(tree=root, minimum=float("inf"), prev=None)
        return res


print()

File: easy/783. Minimum Distance Between BST Nodes/783. Minimum Distance Between BST Nodes.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function getMinimumDifference(root: TreeNode | null): number {
    function run(tree: TreeNode | null, prev: number | null, minimum: number): [number | null, number] {
        if (!tree)
            return [prev, minimum]

        const [prevLeft, minLeft] = run(tree.left, prev, minimum)

        if (prevLeft !== null)
            minimum = Math.min(minLeft, tree.val - prevLeft)

        const newPrev = tree.val

        return run(tree.right, newPrev, minimum)
    }

    const [_, res] = run(root, null, Infinity)
    return res
};